File: 04-Archive/graxia-legacy/packages/bwcp_protocol/python/envelope.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List, Any, Dict

class DeliverableContract(BaseModel):
    type: str
    format: str
    review_required: bool = True
    acceptance_criteria: List[str]

class EscalationPolicy(BaseModel):
    if_blocked: str = "notify_chief_of_staff"
    if_overdue: str = "notify_ceo"
    if_error: str = "notify_engineer_agent"

class MessageEnvelope(BaseModel):
    message_id: str
    thread_id: str
    mission_id: str
    task_id: str
    parent_task_id: Optional[str] = None
    from_agent: str
    to_agent: str
    message_type: str  # MISSION_CREATED, TASK_BID, etc.
    priority: str = "MEDIUM"
    risk_class: str = "CLASS_1"
    budget_limit_usd: float = 0.0
    deadline_at: datetime
    sla_minutes: int = 60
    tool_scope: List[str] = []
    skill_scope: List[str] = []
    requires_approval: bool = False
    confidence: float = 1.0
    deliverable_contract: Optional[DeliverableContract] = None
    escalation: EscalationPolicy = Field(default_factory=EscalationPolicy)
    correlation_id: str
    trace_id: str
    payload: Dict[str, Any] = {}
    created_at: datetime = Field(default_factory=datetime.utcnow)

File: 04-Archive/graxia-legacy/packages/bwcp_protocol/typescript/envelope.ts
import { z } from 'zod';

export const DeliverableContractSchema = z.object({
  type: z.string(),
  format: z.string(),
  review_required: z.boolean().default(true),
  acceptance_criteria: z.array(z.string()),
});

export const EscalationPolicySchema = z.object({
  if_blocked: z.string().default('notify_chief_of_staff'),
  if_overdue: z.string().default('notify_ceo'),
  if_error: z.string().default('notify_engineer_agent'),
});

export const MessageEnvelopeSchema = z.object({
  message_id: z.string(),
  thread_id: z.string(),
  mission_id: z.string(),
  task_id: z.string(),
  parent_task_id: z.string().optional(),
  from_agent: z.string(),
  to_agent: z.string(),
  message_type: z.string(),
  priority: z.enum(['CRITICAL', 'HIGH', 'MEDIUM', 'LOW']).default('MEDIUM'),
  risk_class: z.enum(['CLASS_0', 'CLASS_1', 'CLASS_2', 'CLASS_3', 'CLASS_4', 'CLASS_5']).default('CLASS_1'),
  budget_limit_usd: z.number().default(0),
  deadline_at: z.string().datetime(),
  sla_minutes: z.number().default(60),
  tool_scope: z.array(z.string()).default([]),
  skill_scope: z.array(z.string()).default([]),
  requires_approval: z.boolean().default(false),
  confidence: z.number().min(0).max(1).default(1),
  deliverable_contract: DeliverableContractSchema.optional(),
  escalation: EscalationPolicySchema,
  correlation_id: z.string(),
  trace_id: z.string(),
  payload: z.record(z.any()).default({}),
  created_at: z.string().datetime().default(() => new Date().toISOString()),
});

export type MessageEnvelope = z.infer<typeof MessageEnvelopeSchema>;
